[Post 2.12] Support managed fields in public tokens, kubeconfigs

andreas-kupries opened the pull request:
## Issue: <!-- link the issue or issues this PR resolves here -->

#51171
 
## Problem

The ext token store currently passes the managed field data of the backing secret, instead of data proper for the ext tokens.
 
## Solution

Implemented code performing mapping from token fields to secret fields, and back.
Core code is sharable with ext kubeconfig.
Only the field mappings are specific to tokens.
 
## Testing
<!-- Note: Confirm if the repro steps in the GitHub issue are valid, if not, please update the issue with accurate repro steps. -->

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->

### Automated Testing
<!-- Ensure there are unit/integration/validation tests added (if possible); describe what cases they cover and do not cover. -->
* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  <!-- 
  Non-exhaustive list of reasons:
    - Lack of the framework capable of testing this fix/change
    - Tight deadlines / critical priority to get fix/change in - !ensure GH issue is logged to add tests!
    - No application logic is modified by this change, e.g. refactoring/cosmetic/non-code/test change
    - Tests implemented in another PR elsewhere - !ensure GH PR link is added!
    - Other (explain)
  Note: Outside of the exceptions above, the "existing tests cover the changes" is very unlikely to be an acceptable reason as the existing tests generally don't cover the logic changes implemented by this PR 
  -->
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

pmatseykanets reviewed (commented):
I think mapping is a bit more complicated and should take lists into account. A field with scalar value `f:` in a backing configmap sometimes translates to a list or map  (`v|i|k:`) in the corresponding imperative resource and vice versa.
On a different note the setup of the proposed API seems to be a bit verbose for what potentially could've been just a `map[string]string`.
At this point I'm not sure how generic it can be. I guess we need to write it for tokens and kubeconfigs without worrying too much making it common and then look back and see what we can extract.

andreas-kupries commented:
> I think mapping is a bit more complicated and should take lists into account.

In the basic testing done during dev I had not come across lists yet. Will have to review where that appears. On first thought nothing comes to mind for tokens. ... For kubeconfig however the cluster list comes to mind now. That will require checking.

> A field with a scalar value `f:` in a backing configmap sometimes translates to a list or map (`v|i|k:`) in the corresponding imperative resource and vice versa.

I now suspect that we will require some kind of pattern matching and placeholders. To generally keep part of the input key and place it somewhere in the result key. Because we cannot list a priori all possible cluster names for translation.

> On a different note the setup of the proposed API seems to be a bit verbose for what potentially could've been just a `map[string]string`.

Will think about that some more.

> At this point I'm not sure how generic it can be. I guess we need to write it for tokens and kubeconfigs without worrying too much making it common and then look back and see what we can extract.

Sure.

andreas-kupries commented:
> > I think mapping is a bit more complicated and should take lists into account.
> 
> In the basic testing done during dev I had not come across lists yet. Will have to review where that appears. On first thought nothing comes to mind for tokens. ... For kubeconfig however the cluster list comes to mind now. That will require checking.
> 
> > A field with a scalar value `f:` in a backing configmap sometimes translates to a list or map (`v|i|k:`) in the corresponding imperative resource and vice versa.
> 

Played a bit with ext kubeconfig resources today.
Used two clusters to have a proper cluster list in the resource.
No list data appeared in the set of managed fields.
I saw map data (`k:`), this however was limited to the meta data owner refs.
That is ok for mapping because all of `.metadata` and below is simply passed through.

If the `... sometimes translates ...` is reproducible I would like to see an example.

> > On a different note the setup of the proposed API seems to be a bit verbose for what potentially could've been just a `map[string]string`.

Simplified today to `map[string]fieldpath.Path`, removed the constructors and inversion helper.
 
> > At this point I'm not sure how generic it can be. I guess we need to write it for tokens and kubeconfigs without worrying too much making it common and then look back and see what we can extract.
> 
> Sure.
